File: apps/mobile/src/lib/api/options.ts
import { apiRequest } from "./base";

export type OptionMarket = "tsx" | "commodities";
export type OptionSide = "call" | "put";

export type OptionSourceStatus = {
  source: string;
  status: "available" | "partial" | "unavailable";
  detail: string | null;
};

export type OptionUniverseItem = {
  symbol: string;
  name: string;
  market: OptionMarket;
  category: string;
  exchange: string;
  provider_symbol: string;
  source_url: string | null;
};

export type OptionUniverseSnapshot = {
  market: OptionMarket | "all";
  items: OptionUniverseItem[];
  source_statuses: OptionSourceStatus[];
  generated_at: string;
  refresh_after_seconds: number;
};

export type OptionContract = {
  symbol: string;
  underlying: string;
  market: OptionMarket;
  contract: string | null;
  side: OptionSide;
  strike: number;
  expiration: string;
  exchange: string;
  bid: number | null;
  ask: number | null;
  last: number | null;
  change: number | null;
  percent_change: number | null;
  volume: number | null;
  open_interest: number | null;
  implied_volatility: number | null;
  delta: number | null;
  gamma: number | null;
  theta: number | null;
  vega: number | null;
  in_the_money: boolean | null;
  source: string;
  delayed: boolean;
};

export type OptionChainSnapshot = {
  market: OptionMarket;
  symbol: string;
  provider_symbol: string;
  name: string;
  category: string;
  exchange: string;
  underlying_price: number | null;
  requested_expiration: string | null;
  expirations: string[];
  contracts: OptionContract[];
  analytics: {
    contract_count: number;
    call_count: number;
    put_count: number;
    call_volume: number;
    put_volume: number;
    call_open_interest: number;
    put_open_interest: number;
    put_call_volume_ratio: number | null;
    put_call_open_interest_ratio: number | null;
    atm_implied_volatility: number | null;
    max_pain_estimate: number | null;
  };
  source_statuses: OptionSourceStatus[];
  generated_at: string;
  refresh_after_seconds: number;
  stale: boolean;
};

export const optionsApi = {
  universe: (market: OptionMarket, signal?: AbortSignal) =>
    apiRequest<OptionUniverseSnapshot>(
      `/api/v1/options/universe?market=${market}`,
      { timeoutMs: 15_000, signal },
    ),
  chain: (
    market: OptionMarket,
    symbol: string,
    expiration?: string,
    signal?: AbortSignal,
  ) => {
    const query = new URLSearchParams({ market, symbol });
    if (expiration) query.set("expiration", expiration);
    return apiRequest<OptionChainSnapshot>(
      `/api/v1/options/chain?${query.toString()}`,
      { timeoutMs: market === "commodities" ? 20_000 : 15_000, signal },
    );
  },
};

File: apps/api/app/schemas/options.py
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field


OptionMarket = Literal["tsx", "commodities"]
OptionSide = Literal["call", "put"]


class OptionSourceStatus(BaseModel):
    source: str
    status: Literal["available", "partial", "unavailable"]
    detail: str | None = None


class OptionUniverseItem(BaseModel):
    symbol: str
    name: str
    market: OptionMarket
    category: str
    exchange: str
    provider_symbol: str
    source_url: str | None = None


class OptionUniverseSnapshot(BaseModel):
    market: Literal["tsx", "commodities", "all"]
    items: list[OptionUniverseItem]
    source_statuses: list[OptionSourceStatus]
    generated_at: datetime
    refresh_after_seconds: int = 21_600


class OptionContract(BaseModel):
    symbol: str
    underlying: str
    market: OptionMarket
    contract: str | None = None
    side: OptionSide
    strike: float
    expiration: str
    exchange: str
    bid: float | None = None
    ask: float | None = None
    last: float | None = None
    change: float | None = None
    percent_change: float | None = None
    volume: int | None = None
    open_interest: int | None = None
    implied_volatility: float | None = None
    delta: float | None = None
    gamma: float | None = None
    theta: float | None = None
    vega: float | None = None
    in_the_money: bool | None = None
    source: str
    delayed: bool = True


class OptionAnalytics(BaseModel):
    contract_count: int = 0
    call_count: int = 0
    put_count: int = 0
    call_volume: int = 0
    put_volume: int = 0
    call_open_interest: int = 0
    put_open_interest: int = 0
    put_call_volume_ratio: float | None = None
    put_call_open_interest_ratio: float | None = None
    atm_implied_volatility: float | None = None
    max_pain_estimate: float | None = None


class OptionChainSnapshot(BaseModel):
    market: OptionMarket
    symbol: str
    provider_symbol: str
    name: str
    category: str
    exchange: str
    underlying_price: float | None = None
    requested_expiration: str | None = None
    expirations: list[str] = Field(default_factory=list)
    contracts: list[OptionContract] = Field(default_factory=list)
    analytics: OptionAnalytics = Field(default_factory=OptionAnalytics)
    source_statuses: list[OptionSourceStatus] = Field(default_factory=list)
    generated_at: datetime
    refresh_after_seconds: int = 30
    stale: bool = False